yarn-project/simulator/src/avm/journal/journal.ts

```diff
@@ -94,11 +95,26 @@ export class AvmPersistableStateManager {
     return Promise.resolve(value);
   }
 
+  /**
+   * Check if a note hash exists at the given leaf index, trace the check.
+   *
+   * @param storageAddress - the address of the contract whose storage is being read from
+   * @param noteHash - the unsiloed note hash being checked
+   * @param leafIndex - the leaf index being checked
+   * @returns true if the note hash exists at the given leaf index, false otherwise
+   */
+  public async checkNoteHashExists(storageAddress: Fr, noteHash: Fr, leafIndex: Fr): Promise<boolean> {
+    const gotLeafIndex = await this.hostStorage.commitmentsDb.getCommitmentIndex(noteHash);
```

Comment:
I think this is fine for now, but please add a TODO for siloing by storageAddress.

As you suspected, `noteHash` needs to be siloed by `storageAddress` before it is passed to `getCommitmentIndex`. But technically today in private land, user code checks for existence of a noteHash by outputting the tree leaf itself (the unique siloed note hash). In other words, the protocol doesn't yet enforce that you can only check existence of noteHashes belonging  to the current contract (contract A can check existence of notes from contract B).

So for now, let's just assume that siloing is done by the app. Eventually ([after certain changes in private](https://aztecprotocol.slack.com/archives/C03P17YHVK8/p1708019341409199?thread_ts=1708014666.422409&cid=C03P17YHVK8)) we'll update this function to assume that user-code provided an unsiloed noteHash.